contracts/MarginBase.sol

```diff
@@ -412,4 +454,148 @@ contract MarginBase is MinimalProxyable {
+            return price >= order.desiredPrice;
+        } else {
+            revert("Order size 0");
+        }
+    }
+
+    /// @notice register a limit order internally and with gelato
+    /// @param _marketKey: synthetix futures market id/key
+    /// @param _marginDelta: amount of margin (in sUSD) to deposit or withdraw
+    /// @param _sizeDelta: denoted in market currency (i.e. ETH, BTC, etc), size of futures position
+    /// @param _limitPrice: expected limit order price
+    /// @return orderId contract interface
+    function placeOrder(
+        bytes32 _marketKey,
+        int256 _marginDelta,
+        int256 _sizeDelta,
+        uint256 _limitPrice
+    ) external onlyOwner returns (uint256) {
+        // if more margin is desired on the position we must commit the margin
+        if (_marginDelta > 0) {
+            // ensure margin doesn't exceed max
+            if (_abs(_marginDelta) > freeMargin())
+                revert InsufficientFreeMargin(freeMargin(), _abs(_marginDelta));
+            committedMargin += _abs(_marginDelta);
+        }
+
+        bytes32 taskId = IOps(ops).createTask(
+            address(this),
+            this.executeOrder.selector,
+            address(this),
+            abi.encodeWithSelector(this.checker.selector, orderId)
+        );
+
+        orders[orderId] = Order({
+            marketKey: _marketKey,
+            marginDelta: _marginDelta,
+            sizeDelta: _sizeDelta,
+            desiredPrice: _limitPrice,
+            gelatoTaskId: taskId
+        });
+
+        return orderId++;
+    }
+
+    /// @notice cancel a gelato queued order
+    /// @param _orderId: key for an active order
+    function cancelOrder(uint256 _orderId) external onlyOwner {
+        Order memory order = orders[_orderId];
+
+        // if margin was committed, free it
+        if (order.marginDelta > 0) {
+            committedMargin -= _abs(order.marginDelta);
+        }
+        IOps(ops).cancelTask(order.gelatoTaskId);
+
+        // delete order from orders
+        delete orders[_orderId];
+    }
+
+    /// @notice execute a gelato queued order
+    /// @notice only keepers can trigger this function
+    /// @param _orderId: key for an active order
+    function executeOrder(uint256 _orderId) external onlyOps {
+        require(validOrder(_orderId), "Order not ready for execution");
+        Order memory order = orders[_orderId];
+
+        // if margin was committed, free it
+        if (order.marginDelta > 0) {
+            committedMargin -= _abs(order.marginDelta);
+        }
+
+        // prep new position
+        MarginBase.UpdateMarketPositionSpec[]
+            memory newPositions = new MarginBase.UpdateMarketPositionSpec[](1);
+        newPositions[0] = MarginBase.UpdateMarketPositionSpec(
+            order.marketKey,
+            order.marginDelta,
+            order.sizeDelta,
+            false // assume the position will be closed if the limit order is the opposite size
```

Comment:
Could adjust order struct to include this so if users want to fully close and exit a position at a certain price, then everything can be ignored... freeing up that margin from `commitedMargin`, in that case, would need some extra work though. your call

Reply:
Decided against doing this because it would add another case for keeping track of committed margin